pgxn/neon_walredo/walredoproc.c

```diff
@@ -140,9 +140,42 @@ static XLogReaderState *reader_state;
 #define TRACE DEBUG5
 
 #ifdef HAVE_LIBSECCOMP
+
+
+/*
+ * https://man7.org/linux/man-pages/man2/close_range.2.html
+ *
+ * The `close_range` syscall is available as of Linux 5.9.
+ *
+ * The `close_range` libc wrapper is only available in glibc >= 2.34.
+ * Debian Bullseye ships a libc package based on glibc 2.31.
+ * => write the wrapper ourselves, using the syscall number from the kernel headers.
+ *
+ * If the Linux uAPI headers don't define the system call number,
+ * fail the build deliberately rather than ifdef'ing it to ENOSYS.
+ * We prefer a compile time over a runtime error for walredo.
+ */
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <errno.h>
+int close_range(unsigned int start_fd, unsigned int count, unsigned int flags) {
+    return syscall(__NR_close_range, start_fd, count, flags);
+}
+
 static void
 enter_seccomp_mode(void)
 {
+
+	/*
+	 * The pageserver process relies on us to close all the file descriptors
+	 * it potentially leaked to us, _before_ we start processing potentially dangerous
+	 * wal records. See the comment in the Rust code that launches this process.
+	 */
+	int err;
+	if (err = close_range(3, ~0U, 0)) {
+		ereport(FATAL, (errcode(ERRCODE_SYSTEM_ERROR), errmsg("seccomp: could not close files >= fd 3")));
+	}
+
```

Comment:
This doesn't require seccomp; let's do this even when built without seccomp.

Reply:
I (ab)used the HAVE_SECCOMP as a proxy for that we're on Linux.

Proximity to the seccomp code makes sense IMO because both (and only both) are the essential security building block for walredo.

I'd appreciate if you could provide a follow-up fix that

1. adds a configure check for close_range, so we use the syscall wrapper if it's available in glibc, otherwise fall back to my wrapper & fail to compile if NR_close_range is not defined
2. implements a "status report" walredo command as I suggested on Slack

> Also, I'd like to suggest we add additional robustness checks to walredo, e.g., a message where PS can ask a freshly spawned walredo whether it thinks it correctly set up seccomp.

Reply:
> Also, I'd like to suggest we add additional robustness checks to walredo, e.g., a message where PS can ask a freshly spawned walredo whether it thinks it correctly set up seccomp. 

Is that the "status report" command you meant?

Reply:
Created https://github.com/neondatabase/neon/issues/6580 to track this